### PR TITLE
ARROW-15460: [R] Add as.data.frame.Dataset method

### PR DESCRIPTION
**Pull request by nealrichardson**
Plus some refactoring and disentangling of compute/collect methods

### CODE REVIEW

**Review by wjones127 — CHANGES_REQUESTED** [review]
Changes seem reasonable, but I'm still not super familiar with the metadata stuff. So would like to see that encoded in the unit tests, if possible.

File: r/R/metadata.R [hunk 1]
```diff
@@ -77,7 +77,7 @@ apply_arrow_r_metadata <- function(x, r_metadata) {
         trace <- trace_back()
         # TODO: remove `trace$calls %||% trace$call` once rlang > 0.4.11 is released
         in_dplyr_collect <- any(map_lgl(trace$calls %||% trace$call, function(x) {
-          grepl("collect.arrow_dplyr_query", x, fixed = TRUE)[[1]]
+          grepl("collect\\.([aA]rrow|Dataset)", x)[[1]]
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/R/metadata.R:80]: Yes, a test failed due to other changes and this was the fix.

File: r/R/table.R [hunk 2]
```diff
@@ -330,10 +330,22 @@ as_arrow_table.RecordBatchReader <- function(x, ...) {
   x$read_table()
 }
 
+#' @rdname as_arrow_table
+#' @export
+as_arrow_table.Dataset <- function(x, ...) {
+  Scanner$create(x)$ToTable()
+}
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/R/table.R:337]: Yes good call

**Comment by nealrichardson** [comment]
> Changes seem reasonable, but I'm still not super familiar with the metadata stuff. So would like to see that encoded in the unit tests, if possible.

The schema metadata stuff is already tested, I would guess mostly in test-metadata.R and test-dplyr-group-by.R. This change redirects some codepaths to go through it rather than relying on `arrow_dplyr_query` attributes.

**Pull request merged**